keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ArtifactInterfaces.kt

```diff
@@ -0,0 +1,82 @@
+package com.netflix.spinnaker.keel.api
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact information, typically compute
+ * resources.
+ *
+ * The fields on this interface are nullable because typically the spec will _not_ have that information available
+ * before the corresponding [ResourceHandler] has resolved the resource.
+ */
+interface ArtifactProvider {
+  val artifactName: String?
+  val artifactType: ArtifactType?
+
+  fun completeArtifactOrNull() =
+    if (artifactName != null && artifactType != null) {
+      CompleteArtifact(artifactName!!, artifactType!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain versioned artifacts, typically compute
+ * resources.
+ */
+interface VersionedArtifactProvider : ArtifactProvider {
+  val artifactVersion: String?
+
+  fun completeVersionedArtifactOrNull() =
+    if (artifactName != null && artifactType != null && artifactVersion != null) {
+      CompleteVersionedArtifact(artifactName!!, artifactType!!, artifactVersion!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact references, typically compute
+ * resources.
+ */
+interface ArtifactReferenceProvider {
+  val artifactReference: String?
+  val artifactType: ArtifactType?
+
+  fun completeArtifactReferenceOrNull() =
+    if (artifactReference != null && artifactType != null) {
+      CompleteArtifactReference(artifactReference!!, artifactType!!)
+    } else {
+      null
+    }
+}
+
+interface ComputeResourceSpec : ResourceSpec, VersionedArtifactProvider, ArtifactReferenceProvider
+
+/**
+ * Simple container of the information defined in [ArtifactProvider] that ensures non-nullability of the fields.
+ */
+data class CompleteArtifact(
+  override val artifactName: String,
+  override val artifactType: ArtifactType
+) : ArtifactProvider
+
+/**
+ * Simple container of the information defined by [VersionedArtifactProvider] that ensures non-nullability of the
+ * fields.
+ */
+data class CompleteVersionedArtifact(
+  override val artifactName: String,
+  override val artifactType: ArtifactType,
+  override val artifactVersion: String
+) : VersionedArtifactProvider
+
+/**
+ * Simple container of the information defined by [ArtifactReferenceProvider] that ensures non-nullability of the
+ * fields.
+ */
+data class CompleteArtifactReference(
```

Comment:
can you explain why do we need the differentiation between `CompleteArtifactReference` and `ArtifactReferenceProvider` (for the other classes as well) ?

Reply:
Yep, I added a comment on exactly that here: https://github.com/spinnaker/keel/pull/905#discussion_r395799425